### PR TITLE
Added k-nn user guide and samples.

### PR DESCRIPTION
**Pull request by dblock**
### Description

Added k-nn user guide and samples.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
@vamshin ++ take a look pls?

File: guides/plugins/knn.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        "knn": {
+            "values": {
+                "vector": vec, 
+                "k": 3
+            }
+        }
+    }
+}
+
+results = client.search(index=index_name, body=search_query)
```
Comment (navneet1v) [inline guides/plugins/knn.md:89]: can we add that if we disable the _source field than it helps in reducing the latency of the search.

File: guides/plugins/knn.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+```bash
+$ poetry run knn/knn-boolean-filter.py 
+
+Searching for [0.08, 0.42, 0.04, 0.76, 0.41] with the 'romance' genre ...
+
+{'_index': 'my-index', '_id': '445', '_score': 0.95886475, '_source': {'values': [0.2, 0.54, 0.08, 0.87, 0.43], 'metadata': {'genre': 'romance'}}}
+{'_index': 'my-index', '_id': '2816', '_score': 0.95256233, '_source': {'values': [0.22, 0.36, 0.01, 0.75, 0.57], 'metadata': {'genre': 'romance'}}}
+```
+
+## Approximate k-NN with a Lucene Filter
```
Comment (navneet1v) [inline guides/plugins/knn.md:107]: I know we named it Lucene Filters, but its actually Efficient Filters. We recently changed this in the documentation.

Comment (navneet1v) [reply guides/plugins/knn.md:107]: https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/

File: samples/knn/knn-lucene-filter.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            },
+            "mappings":{
+                "properties": {
+                    "location": {
+                        "type": "knn_vector", 
+                        "dimension": 2,
+                        "method": {
+                            "name": "hnsw",
+                            "space_type": "l2",
+                            "engine": "lucene",
```
Comment (navneet1v) [inline samples/knn/knn-lucene-filter.py:51]: the same filter with 2.9 launch is supported with Faiss also, lets add that example too. Also, lets rename the file to efficient filters.

Comment (dblock) [reply samples/knn/knn-lucene-filter.py:51]: I renamed the file.

File: guides/plugins/knn.md (hunk 1 above)
Comment (dblock) [reply guides/plugins/knn.md:89]: we could, but feels like that's a very narrow use-case where I don't actually have any additional data and only want the ID, no?

**Review by vamshin — APPROVED** [review]
LGTM! Thanks

**Review by kkmr — APPROVED** [review]

**Comment by dblock** [comment]
@VachaShah when you get a chance review/merge pls?

File: guides/plugins/knn.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
+- [k-NN Plugin](#k-nn-plugin)
+  - [Basic Approximate k-NN](#basic-approximate-k-nn)
+    - [Create an Index](#create-an-index)
+    - [Index Vectors](#index-vectors)
+    - [Search for Nearest Neighbors](#search-for-nearest-neighbors)
+  - [Approximate k-NN with a Boolean Filter](#approximate-k-nn-with-a-boolean-filter)
+  - [Approximate k-NN with a Lucene Filter](#approximate-k-nn-with-a-lucene-filter)
```
Comment (Xtansia) [inline guides/plugins/knn.md:7]:
```suggestion
  - [Approximate k-NN with an Efficient Filter](#approximate-k-nn-with-an-efficient-filter)
```

**Review by Xtansia — CHANGES_REQUESTED** [review]
Just a minor issue of a missed rename about lucene filter vs efficient filter.

**Pull request merged**